src/api/app/models/event/request.rb

```diff
@@ -77,16 +77,48 @@ def source_maintainers
       action_maintainers('sourceproject', 'sourcepackage')
     end
 
+    def source_watchers
+      source_or_target_project_watchers(project_type: 'sourceproject')
+    end
+
     def target_watchers
-      find_watchers('targetproject')
+      source_or_target_project_watchers(project_type: 'targetproject')
     end
 
-    def source_watchers
-      find_watchers('sourceproject')
+    def source_package_watchers
+      source_or_target_package_watchers(project_type: 'sourceproject', package_type: 'sourcepackage')
+    end
+
+    def target_package_watchers
+      source_or_target_package_watchers(project_type: 'targetproject', package_type: 'targetpackage')
     end
 
     private
 
+    def source_or_target_project_watchers(project_type:)
+      # TODO: remove old `find_watchers` implementation after rolling out `new_watchlist` from beta program
+      find_watchers(project_type) +
+        payload['actions'].map { |action| action[project_type] }
+                          .map { |project_name| Project.find_by_name(project_name) }
+                          .compact.map(&:watched_items)
+                          .flatten.map(&:user)
```

Comment:
Shouldn't we add here a `rescue Project::Errors::UnknownObjectError`?

Reply:
Not this one, the `Project.find_by_name` will return `nil` instead of :boom:, so we don't need to catch anything. Then `nil` will be discarded by the `compact` call.

Reply:
Ops, true. I got confused with `Project.get_by_name`.

By the way, `Project.find_by_name` [will be obsoleted](https://github.com/openSUSE/open-build-service/blob/4842f9f1c85a1b5b9050d2f0025fd4b215846e10/src/api/app/models/project.rb#L259-L261). We can live with it. Having around 40 ocurrences of Project.find_by_name in the code it is a signal that we still should use this method, and that at least one PR only refactoring these calls would be needed.